### PR TITLE
Refactor: Move matrix_box_presenter.rb#get_rss_log_details => rss_log.rb#detail

### PR DESCRIPTION
**Pull request by nimmolo**
Delivers a longstanding TODO, refactoring an rss log presenter method to the model
No **intended** change in current functionality, simply a refactor. (But please check thoroughly!)

Needed for current GraphQL API work - makes RssLog details available directly

### CODE REVIEW

**Comment by nimmolo** [comment]
Hi everyone -
I just noticed, after isolating and testing this on my graphql branch, that the `detail`/`get_rss_log_detals` method here initiates secondary queries on the User table for every RssLog that refers to an Observation or Species List. 

That's the case on the original method too. These queries are not batched and pretty costly... but obviously the user in question is an essential element of the RssLog itself. 

Can anyone think of a way to refactor -- maybe set up a "belongs to" association that would grab and store a user_id in the RssLogs table... That way at least we can batch load the association. Just thinking out loud here.

**Comment by pellaea** [comment]
The way I've been dealing with this is to ensure that everyone who uses
get_rss_log_details eager-loads the user.  For example, all observation
indexes need this.  See line 120 in
app/controllers/observer_controller/indexes.rb.  This is where
it specifies which relations to eager-load for all the observation
indexes.  But note that it is already eager-loading the user there.
Another place it does it is around line 50 of
app/controllers/observer_controller/rss_log_controller.rb, but it is
eager-loading the user for each possible target type there, too.

Okay, time to track down all uses of get_rss_log_details, which is in
MatrixBoxPresenter. MatrixBoxPresenter is only used by shared/matrix_box
view, which in turn is used by:

app/views/image/list_images.html.erb
app/views/herbarium_record/show_herbarium_record.html.erb
app/views/herbarium_record/edit_herbarium_record.html.erb
app/views/herbarium_record/create_herbarium_record.html.erb
app/views/observer/list_rss_logs.html.erb
app/views/observer/list_users.html.erb
app/views/observer/list_observations.html.erb
app/views/collection_number/show_collection_number.html.erb
app/views/collection_number/edit_collection_number.html.erb
app/views/collection_number/create_collection_number.html.erb

HerbariumRecordController and CollectionNumberController, are definitely
not doing any eager-loading, and they need to.  But the others all look
okay.  It might be worth double-checking by watching the queries in
development.log while running the server locally.

-Jason

**Comment by pellaea** [comment]
Andrew, I see the problem.  The way you've refactored it will break the
eager-loading.  It's important to understand the full linear execution flow
in cases like this (which admittedly object oriented programming makes as
difficult as it possibly can).  Take the example of an observation index.

1. Query loads a bunch of observations, eager-loading the user for each one:

@observations = Observation.where(...).include(:user)

This means that if you access the user of an individual observation like
this:

@observations[n].user.login

It will already be loaded, and no extra DB hits required.

2. Now it sends that list of observations to the view, then it sends each
observation (and its rss_log) to MatrixBoxPresenter, where it access the
log details via observation.rss_log.detail (line 89 of
app/presenters/matrix_box_presenter.rb).

3. However, RssLog#detail doesn't have access to the original observation,
so it now has to go backwards to get back to the observation.  In effect it
does this:

@observations[n].rss_log.observation.user.login

That is not eager-loaded.  I recommend that you allow the caller of
RssLog#Detail to supply the target explicitly -- **even though the rss log
clearly already has access to the target!**  That is, allow
MatrixBoxPresenter to give it the target explicitly:

rss_log.detail(target)

and have RssLog#detail access user via the passed-in target if available.
In fact, you may even *require* the caller to pass in the target
explicitly, in order to avoid this problem in the future.  If it's
unnecessary the caller can always do this:

rss_log.detail(rss_log.target)

But it forces the caller at least to think "wait, will the target (and
target's user) be eager-loaded correctly, or is this going to result in a
hundred expensive little DB queries?"

(A more railsy solution might be to override RssLog#target, and have it
default to the original, but allow the caller to explicitly set the target
with their own instance of the target if eager-loading is an issue.  But
this sounds like compounding the problem: Rails makes it virtually
impossible to determine implementation details, making optimization like
this a real bear.  Is overriding an automagical Rails method really the
right solution to overuse of automagic??)

I hope I managed to explain that clearly!

-Jason

**Comment by pellaea** [comment]
PS. Otherwise this refactor is a great idea and I think the way you did it
is just fine.

**Comment by nimmolo** [comment]
@pellaea Thanks for these clear explanations and all the consideration that went into them. I woke up thinking the same thing re: how calling it from the presenter allowed for automagical eager loading and this breaks it. 

I like your solution of calling it with (target) for the use of the current matrix box presenter - however i don't know if this will help with graphql, which wants to resolve each line item in the query on its own terms. 

In a graphql query for Observations, for example, as far as I know, i can't pass the id of a parent observation down to the nested rsslog model. (I'm going to do some digging in graphql-ruby today and try to find a way to reference the parent object being iterated. That way I could pass it as "target". But i'm not sure it's possible.)

```graphql
query getObservations {
  observations {
    edges {
      cursor
      node {
        id
        thumbImageId
        textName
        name {
          id
          author
          displayName
          formatName
        }
        user {
          id
          name
          login
        }
        where
        when
        createdAt
        rssLog {
          detail    <= this simply fires the method and doesn't know anything about the current iteration?
          parseLog
          notes
        }
      }
    }
    pageInfo {
      startCursor
      endCursor
      hasNextPage
      hasPreviousPage
    }
  }
}
```

**Comment by nimmolo** [comment]
I'm now noticing however - the legal name is right there in the notes being parsed. Here's Christian Schwarz changing the consensus on Nathan's bolete:
```graphql
        {
          "cursor": "MTA",
          "node": {
            "id": 10,
            "thumbImageId": 11,
            "textName": "Xerocomellus dryophilus",
            "name": {
              "id": 43604,
              "author": "(Thiers) N. Siegel, C.F. Schwarz & J.L. Frank",
              "displayName": "**__Xerocomellus dryophilus__** (Thiers) N. Siegel, C.F. Schwarz & J.L. Frank",
              "formatName": "<b><i>Xerocomellus dryophilus</i></b>"
            },
            "user": {
              "id": 1,
              "name": "Nathan Wilson",
              "login": "nathan"
            },
            "where": "Chumash Interpretive Center, Thousand Oaks, Ventura Co., California, USA",
            "when": "2002-01-05",
            "createdAt": "2006-05-21T03:17:22-04:00",
            "rssLog": {
              "detail": "Consensus changed from: <i>Xerocomus dryophilus</i> (Thiers) Singer",
              "parseLog": "[[:log_consensus_changed, {:new=>\"**__Xerocomellus dryophilus__** (Thiers) N. Siegel, C.F. Schwarz & J.L. Frank\", :old=>\"__Xerocomus dryophilus__ (Thiers) Singer\", :user=>\"Christian Schwarz\"}, Wed, 23 Jul 2014 13:59:03 EDT -04:00], [:log_naming_created_at, {:name=>\"**__Xerocomellus dryophilus__** (Thiers) N. Siegel, C.F. Schwarz & J.L. Frank\", :user=>\"Christian Schwarz\"}, Wed, 23 Jul 2014 13:58:48 EDT -04:00]]",
              "notes": "20140723175903 log_consensus_changed new **__Xerocomellus%20dryophilus__**%20(Thiers)%20N.%20Siegel,%20C.F.%20Schwarz%20&%20J.L.%20Frank old __Xerocomus%20dryophilus__%20(Thiers)%20Singer user Christian%20Schwarz\n20140723175848 log_naming_created_at name **__Xerocomellus%20dryophilus__**%20(Thiers)%20N.%20Siegel,%20C.F.%20Schwarz%20&%20J.L.%20Frank user Christian%20Schwarz\n"
            }
          }
        },
```
So the query is just to find out if the username is redundant...It's polite but not strictly necessary?
```ruby
      unless [:observation, :species_list].include?(target_type)
        begin
          result += " ".html_safe + :rss_by.t(user: target.user.legal_name)
        rescue StandardError
          nil
        end
      end
    else
      if [:observation, :species_list].include?(target_type) &&
         [target.user.login, target.user.name, target.user.legal_name].
         include?(args[:user])
        # This will remove redundant user from observation logs.
        tag2 = :"#{tag}0"
```

What if I wrote a tailored method for Observation `detail_obs` that simply parses the log and takes the most recent, assembling a plain text note?

**Comment by pellaea** [comment]
Careful.  The owner of the observation may not be the same as the user who,
say, made a comment or proposed a name on that observation.

Point taken about graphql wanting to do this on its own terms.  But if
that's so, then there is probably already a solution.

What about this: rss_log.detail should only be reporting rss_log-y stuff,
not observation-y stuff.  That means it should be reporting the user who
made the comment, not the user who owns the target (observation).  Can we
separate out these things so that calling methods on observations only
returns stuff immediately available in the observation, and calling methods
on rss_logs only returns stuff immediately available in the rss_log?

Then problem totally solved.

-Jason

PPS. I want to thank you for noticing this!  I was just about to review the
PR and I'm positive I would not have noticed that eager-loading was about
to get broken.

**Comment by pellaea** [comment]
Thinking about this more: I think this means that the presenter should be
calling "detail" not on the rss_log, but on the observation.  The
observation, in turn, asks the rss_log to give the latest activity.  Seems
to me the rss_log should not be responsible for building the search result
for an observation (or name, location, etc. etc.), but rather just for
reporting *activity* on those various objects.  Does that make sense?

Again, I think with this different approach it makes everything work more
smoothly.  The fact that we needed to do:

observation.rss_log.target.blah

should have been a big red flag that we'd designed something wrong.

**Comment by nimmolo** [comment]
Yes! This is what i'd like to do. 

This method currently retrieves the "rsslog action instigator" (commenter, proposer of name) by parsing the first note, and checks that against the target.user.legal_name (user who created the target) to avoid duplication.  

Instead, the presenter can handle that user name check. The details method would return the action instigator's name eg "Christian Schwarz" in all cases, and the presenter could print it or no. As could graphQL. 

So: should `details` return a hash of the two string segments, e.g. {action: "Name proposed: Xylaria zoomorpha" by: "Christian Schwarz"} or should it concatenate and let the presenter deal with a "Name proposed: Xylaria zoomorpha, by Christian Schwarz" string?

**Comment by nimmolo** [comment]
Got it. Makes sense... So each model that gets an rsslog would have its own detail method?

**Comment by pellaea** [comment]
Your solution probably makes more sense.  Your solution also removes the
need for the rss_log to know anything about its target.  Probably simpler
and less changing existing code.

My two cents: gave RssLog#detail return a hash.  I feel like RssLog
shouldn't have to worry about how to display its data, just provide access
to it.

Glad we're on the same page!

-Jason

**Comment by nimmolo** [comment]
OK. I'm not sure i understand the method entirely. 
If it's obvious how to do it feel free to do a quick commit, i'll pull it down and handle testing etc.

What i'm noticing is the detail itself (the item in the notes array) is a three-item array structured like this
```ruby
[:log_consensus_changed, 
 {:new=>\"**__Xerocomellus dryophilus__** (Thiers) N. Siegel, C.F. Schwarz & J.L. Frank\", 
  :old=>\"__Xerocomus dryophilus__ (Thiers) Singer\", 
  :user=>\"Christian Schwarz\"}, 
 Wed, 23 Jul 2014 13:59:03 EDT -04:00]
```
and pulled like this in the method
``` ruby
      tag, args, time = parse_log.first
```
But i don't understand how `RssLog#detail` handles the args separately - this is some ninja Ruby

**Comment by nimmolo** [comment]
It seems like maybe `detail` should return a THREE part hash - basically just the results of `parse_log.first` translated into plain... uh, English, or Spanish, or Russian... via the appropriate translated strings.

**Comment by nimmolo** [comment]
OK. Here's what i've got, deleted parts commented out. (Is there a Ruby way to run this method in isolation, without running it through the presenter?) Hash would be result = {notice: notice by: by}
```ruby
  # Figure out a detail message for most recent update.
  def detail
    target_type = target ? target.type_tag : target_type
    begin
      tag, args, time = parse_log.first
      by = args[:user] ? :rss_by.t(args[:user]) : nil
    rescue StandardError
      []
    end
    if !target_type
      notice = :rss_destroyed.t(type: :object)
    elsif !target ||
          tag.to_s.match(/^log_#{target_type}_(merged|destroyed)/)
          notice = :rss_destroyed.t(type: target_type)
    elsif !time || time < target.created_at + 1.minute
      notice = :rss_created_at.t(type: target_type)
      # unless [:observation, :species_list].include?(target_type)
      #   begin
      #     result += " ".html_safe + :rss_by.t(user: target.user.legal_name)
      #   rescue StandardError
      #     nil
      #   end
      # end
    else
      # if [:observation, :species_list].include?(target_type) &&
      #    [target.user.login, target.user.name, target.user.legal_name].
      #    include?(args[:user])
      #   # This will remove redundant user from observation logs.
      #   tag2 = :"#{tag}0"
      #   result = tag2.t(args) if tag2.has_translation?
      # end
      unless notice
        tag2 = tag.to_s.sub(/^log/, "rss").to_sym
        notice = tag2.t(args) if tag2.has_translation?
      end
      begin
        notice ||= tag.t(args)
        result = {notice: notice, by: by}
      rescue StandardError
        nil
      end
    end
    result
  end
```

**Comment by pellaea** [comment]
This looks right to me, I think you've done a good job walking the line
between minimizing required changes while moving the code in the right
direction and making it cleaner.  You're right, this is some gnarly code!
Backwards compatibility is the bane of clean code! :)

**Comment by nimmolo** [comment]
Thanks. I’m checking the return values now…
Idk if it’s the presenter or the method that’s borgled, but the presenter is not always printing the actual detail message. Sometimes only a byline.

Will report back in a few.

Nimmo

**Comment by pellaea** [comment]
BTW, as far as I'm concerned you do not need to sweat the ABC complexity
and other metrics -- it's a pre-existing condition.  Your work is already
extremely valuable.  You don't need to fix every last problem you notice
along the way! :)

File: app/models/rss_log.rb
Comment (nimmolo) [reply app/models/rss_log.rb:383]: I'm guessing the preferred syntax would be the following, but i'm not too familiar with `match?`
```ruby
elsif tag.to_s.match?('/^log_#{target_type}_(merged|destroyed)/')
```
Let me know if that seems right and i'll add this commit.

**Comment by nimmolo** [comment]
I believe this PR is good to go, in that CodeClimate's problems with the file seem minor. 
You guys may be able have input on the `match?` vs `~=` line, see comment above.

Otherwise, I believe I've addressed @pellaea 's improvement suggestions and concerns, and would like to suggest this PR can be merged if it looks good to everybody.

**Comment by pellaea** [comment]
Joe is probably the authority on this, but I do seem to remember that
rubocop prefers your alternative.
Sorry I never found time to review this PR.  I'm glad you're getting it out
finally.

**Comment by nimmolo** [comment]
Jason - you reviewed it very thoroughly and caught a bunch of gotchas. I appreciate it!

File: app/models/rss_log.rb
Comment (nimmolo) [reply app/models/rss_log.rb:383]: I believe the regex in the parens should not have single quotes, per [APIDock](https://apidock.com/ruby/String/match). 
```ruby
elsif tag.to_s.match?(/^log_#{target_type}_(merged|destroyed)/)
```

**Comment by nimmolo** [comment]
Should I write a test for this function specifically? What would it be testing, if so?

File: app/models/rss_log.rb
Comment (nimmolo) [reply app/models/rss_log.rb]: We were previously returning this in the hash, but turns out the user is in the args hash already, so `by` is unnecessary. The whole method is pretty simple now.

File: app/models/rss_log.rb
Comment (nimmolo) [reply app/models/rss_log.rb]: Done

File: app/models/rss_log.rb
Comment (nimmolo) [reply app/models/rss_log.rb]: Done

File: app/presenters/matrix_box_presenter.rb
Comment (pellaea) [inline app/presenters/matrix_box_presenter.rb:56]: Can we do this without calling rss_log.detail twice?  Maybe something like this:

    return unless temp = rss_log.detail
    self.detail = temp

File: config/locales/en.txt
Comment (pellaea) [inline config/locales/en.txt:1094]: Do you remember why [new] is being removed?

File: config/locales/en.txt
Comment (pellaea) [inline config/locales/en.txt:1944]: Do you remember why these are all being removed?  Are these no longer produced anywhere?  Have you checked that they are not present in the existing database, too?

**Comment by pellaea** [comment]
> Should I write a test for this function specifically? What would it be testing, i

Err, which function needs testing?  RssLog#detail?

File: config/locales/en.txt
Comment (nimmolo) [reply config/locales/en.txt:1094]: That may have been my own flourish, it's starting to ring a bell.

Is it reliable assumption that for _a log of an observation name change_, [new] == the title of the log item? (If a log is generated every time consensus changes...) In that case i felt it was less verbose, but it seems like I would have run that by you at the time 🤔 

Or is that an unreliable assumption, and the title of the log will always be the current (as of right now) consensus — so if the consensus changes three times, it's not accurately showing from what to what each time...?

File: config/locales/en.txt
Comment (nimmolo) [reply config/locales/en.txt:1944]: For all the removals, these came from an email you sent me. I triple checked them before committing the deletes, but that doesn't mean i didn't mess up. Will post your list here as soon as I find it.

**Comment by nimmolo** [comment]
Right, that's what i was wondering, if rss_log#detail needed a test. Coveralls said coverage went down.

File: app/presenters/matrix_box_presenter.rb
Comment (nimmolo) [reply app/presenters/matrix_box_presenter.rb]: Just committed this change.

File: config/locales/en.txt
Comment (nimmolo) [reply config/locales/en.txt:1944]: @pellaea This was your proposed list:

> Change these:
>
>   # log_consensus_changed: "Consensus changed from: [old]" <--- maybe not, see below
>   log_naming_created_at: "Name proposed by [user]: [name]"
>   log_naming_destroyed: "Name proposal destroyed by [user]: [name]"
>   log_naming_updated: "Name proposal updated by [user]: [name]"
> 
> Delete these:
> 
>   rss_consensus_changed
>   rss_naming_created_at
>   rss_naming_destroyed
>   rss_naming_updated
>   log_object_added_no_user
>   log_object_created_no_user
>   log_object_destroyed_no_user
>   log_object_merged_no_user
>   log_object_moved_no_user
>   log_object_removed_no_user
>   log_object_updated_no_user
>   log_image_created0
>   log_image_destroyed0
>   log_image_removed0
>   log_image_reused0
>   log_image_updated0

File: config/locales/en.txt
Comment (nimmolo) [reply config/locales/en.txt:1094]: @pellaea This was your thinking on this matter at the time. 

> PS. log_consensus_changed -- This is a tricky one.  I think the motivation was to shorten the matrix box, because the item will already have the latest name, so if you leave it as "Consensus changed from [old] to [new]" the new name will appear twice in the matrix box, and names (esp. with the authorities!) can be huge.  But when you ask to see the log for that observation, it will just show the old name, not the new name.  You see, this override thing was all about allowing us to display something slightly different for the observation / activity index versus the "Show Log" for an observation.  ...   Let's just make them the same, and use the version that looks best on the activity log / observation index page (matrix box).  So that means removing "[new]" from log_concensus_changed.  Make sense?

**Comment by pellaea** [comment]
Great!  Thanks for the reminders.  Impressive.  I can't imagine
successfully looking up those old emails.  Glad you were able to find
them.  I'll defer to my past self.  Since I'm getting noticeably dumber
each year, that seems like the prudent course. :)

re: testing RssLog#detail -- Definitely that should be tested nominally.
There have got to be a few nontrivial rss_logs in the fixtures, right?
Should be easy peasy.

Good work... now that I'm reminded of the reasoning for our
simplifications, I can relax into being excited about the progress it
represents!

-Jason

**Comment by nimmolo** [comment]
Added test. It seems like a bit much, but i'm honestly not sure what an adequate level of coverage is... If someone has an opinion, please suggest shortening it. The second thru 4th assertions seem like the core of it. I don't understand why the notes for the "albion" log are "Updated Created", but that's from the fixture.

**Comment by pellaea** [comment]
Actually, shoot, the fixtures are all fake and useless for tests, nothing
like the logs you would find in real life.  Well, it's good to have one
that just a plain old string, I suppose -- there might still be one or two
of those lurking from the early days of MO.  But it would be good to change
at least a few to be more realistic.  It would be great to create an
orphaned log, too -- either add one to the fixtures, or just create one on
the fly in the test.  Make sure all three basic cases are covered from
RssLog#detail: object deleted, object merged, object modified, ... are
there any other cases that should be tested?  Does this make sense?

**Comment by nimmolo** [comment]
It does make sense. Now, one of the few existing `RssLogTest`s does create an orphaned log. I'm guessing that test method should also be extended to check the detail, within the same test method (because tests tear down after each method, the state doesn't persist).  I don't understand how the `create_rss_log` helper in `RssLogTest`comes into it - it doesn't seem to be used by any of these tests. But it seems like i'd use it here.

Sounds like i also need to make some new tests, for example:

- One that deletes an object, and checks the log detail afterwards
- One that merges an object into another,  and checks the log detail afterwards
- One that modifies an object, and checks the log detail afterwards

But  what objects should I delete/merge/modify?  ...  I'm guessing not the log fixtures themselves, but some underlying object fixtures, like a glossary term. Are other fixtures available to these tests?

i'm realizing i don't know much about Rails testing (to put it mildly).

**Comment by pellaea** [comment]
No worries,  these are all good questions!  The fixtures are actually
restored after each test.  (They implicitly wrap each test in a mysql
transaction and then rewind the transaction after each test -- very elegant
and efficient... and surprisingly it even works. ;)

Your plan to actually delete, merge and modify an object and see what the
rss log detail looks like should work.  And probably there is a camp that
would favor that solution.

Another camp might recommend keeping it simpler.  After all, the point of
this test isn't to test that the hooks for deleting and modifying objects
correctly adjust the associated rss_log.  The point of this test is just to
test that if the rss_log has xxx then it renders the detail like yyy.

Using the second approach, I'm recommending changing, say, the first
rss_log fixture's notes from "Updated Observation & Notes" to:
"20090722075918 log_observation_created user ignatz", then your test would
look like:

  rss_log = rss_logs(:observation_rss_log)
  assert_equal(:log_observation_created.t(user: "ignatz"), rss_log.detail)
  assert_equal(20090722075918, rss_log.time)

(The time comparison won't work that way... probably need to parse the
string "20090722075918" and do a one second interval test or something like
that.  I don't remember the details.  Asserting timestamps are right always
seems to be unnecessarily difficult.)

(You probably also want to check that the format of the rss_log notes is
correct.  (See RssLog#parse for details.  I just grabbed that example from
near app/models/rss_log.rb line 100.)

Also, there is always at least a very slight risk that changing a fixture
will break some other unrelated test.  I think Joe would recommend never
doing that.  But in this case I would make the argument that the fixtures
should actually look like real live objects.  But as currently written, the
rss_log fixtures do not resemble what rss_logs actually look like in the
wild.  So in a sense you would be fixing a long-standing "bug" in our tests.

Just my opinions, of course.  If you are doing the coding, you are welcome
to do it however you like -- whatever makes the most sense to you, or
whatever you have time and interest in doing.

I'm kicking off for the night, now, too.  I'll check back in tomorrow.

File: test/models/rss_log_test.rb
Comment (nimmolo) [reply test/models/rss_log_test.rb:34]: I don't understand how to fix this. 
If i do `Time.zone.parse`, the time is seemingly off by the number of hours offset - in other words, it double-offsets the time zone.

How do i fix this? What is wrong with `Time.parse` anyway?

```
        --- expected
        +++ actual
        @@ -1 +1 @@
        -Wed, 22 Jul 2009 07:59:18 EDT -04:00
        +Wed, 22 Jul 2009 03:59:18 EDT -04:00
```

**Comment by pellaea** [comment]
Yeah, this time zone stuff is frustrating.  I'm right there with you.  Let
me check this out and see if I can see where it's off.  Maybe the fixture
needs to be time zonified, too.  Or maybe the parser isn't zone aware.  (It
wouldn't necessarily be a problem if everything was done in UTM by the same
server; that would hide improper handling of time zones.)

**Comment by pellaea** [comment]
Did you see how RssLog parses it??  Ugh.  RssLog#decode should probably be
using Time.zone.parse... maybe Time.zone("utc").parse??  I don't remember
how it is supposed to work.  Would you mind looking up the docs to learn
how to correctly parse a UTC timestamp?  (And make sure it produces the
same result as that horrble Time.utc construct... then throw it as far as
you can. :)

File: test/models/rss_log_test.rb
Comment (nimmolo) [reply test/models/rss_log_test.rb:34]: OK, `rss_log.decode` was doing it this way:

```ruby
      time2 = Time.utc(time[0, 4], time[4, 2], time[6, 2],
                       time[8, 2], time[10, 2], time[12, 2]).in_time_zone
```
If i change that to 
```ruby
      time2 = Time.zone.parse(time)
```
then expected and actual are the same. 

However, that means they're now both `Wed, 22 Jul 2009 03:59:18 EDT -04:00` which implies `rss_log.decode` was parsing time incorrectly before. 

Which is the correct parse of `20090722075918`? As far as I can make out, there is no time zone encoded in that string, so I assume it's GMT. And what time zone do the tests assume?

**Comment by pellaea** [comment]
Well, it's *encoded* using:

    time.utc.strftime("%Y%m%d%H%M%S")

However you decode it, make sure it is unchanged by a round trip:

    assert(Time.zone.parse("20090722075918").utc.strftime("%Y%m%d%H%M%S")
== "20090722075918")

Does that make sense?

File: test/models/rss_log_test.rb
Comment (nimmolo) [reply test/models/rss_log_test.rb:34]: So, `Time.parse(time).in_time_zone` is the way to go. 

This returns the same parsed time as the more verbose function in `rss_log.decode` currently.

This assertion returns true, for example:
```ruby
assert_equal(Time.parse("20090722075918").in_time_zone.utc.strftime("%Y%m%d%H%M%S"), "20090722075918")
```

**Comment by JoeCohen** [comment]
I doubt I'll have time to review this. 
Why not just merge? (The coverage decrease is trivial.)

**Comment by nimmolo** [comment]
@JoeCohen   It's a little hard to follow, sorry. I responded to Jason's questions on the thread underneath the codeclimate issue, so they seem out of sequence here (if you're checking GitHub.com). 

I believe everything is resolved, and we have a new working test as well as some improved code in the model. 

Let's wait til Jason chimes in though, I was really following his lead here.

**Comment by pellaea** [comment]
I'm with Joe here.  Let's just merge the blasted thing and see how it works live.  It just affects presentation, so it's totally safe in that sense.  It looks like it's getting time stamps and basic messages right.  And it's much better code.  The only risks are that we somehow missed some important cases, or that it will result in some unexpected expensive queries.  We've checked for the latter, but you never know for sure.

At least that's my take on the subject.

**Comment by nimmolo** [comment]
Just collapsed some of my earlier comments (on GitHub's website, doesn't affect the email thread) to make this PR discussion easier to follow. I noticed someone doing that in another PR/another repo and i thought it seemed considerate.

**Comment by JoeCohen** [comment]
I'll merge this and deploy it on the webserver later this morning. (I want to be around in case something goes wrong, but have to go out now until ~09:30 PST.)

**Comment by JoeCohen** [comment]
Back. But I want to manually exercise this a bit more before merging.
It gives the wrong message when an Observation is destroyed.
Master:
>Agaricomycetes Doweld
**Observation Destroyed**
11 minutes ago

In #808 
>Agaricomycetes Doweld
**Agaricomycetes Doweld**
7 minutes ago

I want to see if there are more like this.

**Comment by nimmolo** [comment]
Joe, thanks!! Will revise per whatever you find.

**Comment by nimmolo** [comment]
@pellaea 
The unexpected issue occurs when, for example, a location is merged — `detail` seems to be producing the name of the merged item instead of the proper merged message.

We did some digging in the db — The rss_log.notes column for a location created just now locally, and then merged into another location, reads like this (i’m separating the entries for clarity): 
`USA,%20California,%20Berkeley,%2011%20Spinnaker%20Way,%20César%20E.%20Chávez%20Park`  `20220118200103 log_location_merged that Berkeley%20Marina,%20Berkeley,%20Alameda%20Co.,%20California,%20USA this USA,%20California,%20Berkeley,%2011%20Spinnaker%20Way,%20César%20E.%20Chávez%20Park user nathan`  `20220118195857 log_location_created user nathan`

This is consistent with old merged locations in the db. Here's another example **not** created by me just now.
`winston%20salem,%20North%20Carolina,%20USA` `20210610151117 log_location_merged that Winston-Salem,%20North%20Carolina,%20USA this winston%20salem,%20North%20Carolina,%20USA user MO%20Webmaster` `20210609182258 log_location_created user akamat`

So it seems this is not a new pattern for log entries, but my detail method is mis-parsing the log entry in the notes. It needs to skip that first string (the location name, which is not a proper entry?), and get the first entry with a parseable date, or something like that.

Does that sound right?

**Comment by nimmolo** [comment]
It seems that `parse_log` sends a `tag` of `:log_orphan` and just the `title`, if there’s no timestamp (no block of 14 digits). See here:

```ruby
  def parse_log(cutoff_time = nil)
    first = true
    results = []
    for line in notes.to_s.split("\n")
      if first && !line.match(/^\d{14}/)
        tag  = :log_orphan
        args = { title: self.class.unescape(line) }
        time = updated_at
      elsif line.present?
        tag, args, time = self.class.decode(line)
      end
      break if cutoff_time && time < cutoff_time

      results << [tag, args, time]
      first = false
    end
    results
  end
```
This seems to imply that my `detail` method's check for a `log_#{target_type}_(merged|destroyed)` message within such a tag will always return false:
```ruby
    elsif tag.to_s.match?(/^log_#{target_type}_(merged|destroyed)/)
      :rss_destroyed.t(type: target_type)
```

I understand this is working currently, i just don't understand how.

**Comment by nimmolo** [comment]
OK!! I remember the issue now.

The old way the presenter does it is this:
```ruby
    if !target_type
       self.detail = :rss_destroyed.t(type: :object)
     elsif !target ||
           tag.to_s.match(/^log_#{target_type}_(merged|destroyed)/)
       self.detail = :rss_destroyed.t(type: target_type)
```
The problem is that the call of `!target ||` in the old presenter method makes the merge message appear, but putting it in my model method initiates n+1 queries, thereby defeating one of the original purposes of this refactor.

Let's see if we can think of a way to pass the correct merged message to `detail` without checking that there's not a target. The message is right there in the second line of the log, can that be the one returned by `parse_log` instead?

**Comment by nimmolo** [comment]
I'm restoring `!target ||`. I can't find that it's loading an extra query on current MO. It may be that it initiates an extra query if the method is called via GraphQL, in which case i'll submit another PR.

This commit restores the current behavior for merges and destroys. If we want to change it, we know where to change it, but this PR was not intended to change functionality.

In my opinion, the current `parse_log` returning a "Location Destroyed" message for a merge is not ideal, despite its admirable brevity, because it implies we got rid of "Washington State" when _Jason, Destroyer of Worlds_ actually took time out of his day to merge it into "Washington, USA". It seems to me the latter message might be educational for users who aren't aware that duplicate locations, names, etc are a chore to remove.

**Comment by JoeCohen** [comment]
This looks great.  Thanks for digging through the mess to fix the destroyed Observations and Locations.
- I agree that it should be merged, but will wait to hear from @pellaea and @mo-nathan.
- I also agree that the merger messages are confusing for non-developers and could be improved (in another PR).
- Loved "Jason, Destroyer of Worlds" :wheel_of_dharma: :supervillain_man:. Historical note: before Location Earth was locked, it was possible to destroy it, and that happened several times.

**Comment by pellaea** [comment]
Oops, too many emails too little time! -- is this resolved?  Do I still
need to weigh in on anything?

**Comment by pellaea** [comment]
Urk.  Still hard to follow when my brain is fried and scattered!  I understand that the Activity Log has been restored and now displays "Object destroyed" after an object is destroyed?  Also, I think I'm hearing that it may make more sense to say "Object merged into blah" in some cases.  (That is, it should be taking the next-to-last log message instead of always saying just "object destroyed"?  Am I getting that right?)  I'm all for improving the information content of Activity Log messages.  However, I suspect that this PR has restored the original behavior...  If so, I say -- unless it's trivial to fix -- merge this thing and create a new issue for the (ultra-low-priority) improvement of the Activity Log message.  My two cents!

**Comment by JoeCohen** [comment]
I think that's exactly what Nimmo wants to do.

**Comment by nimmolo** [comment]
Confirm

**Review by mo-nathan — APPROVED** [review]
Looks reasonable to me.  I'm not 100% clear on what the old special cases was around species lists, but the new code looks better to me.  I have not tested it throughly, but I agree with the consensus that this is better and if we find a problem with it we can fix it then.

**Review by JoeCohen — APPROVED** [review]

**Pull request merged**
